File: leetcode/0060-permutation-sequence/0060-permutation-sequence.java
class Solution {
    public String getPermutation(int n, int k) {
        // StringBuilder to build the result permutation
        StringBuilder result = new StringBuilder();
      
        // Track which numbers have been used (1-indexed for convenience)
        boolean[] used = new boolean[n + 1];
      
        // Build the permutation digit by digit
        for (int position = 0; position < n; position++) {
            // Calculate factorial of remaining positions
            // This represents how many permutations each choice contributes
            int factorial = 1;
            for (int i = 1; i < n - position; i++) {
                factorial *= i;
            }
          
            // Try each number from 1 to n
            for (int number = 1; number <= n; number++) {
                // Skip if number is already used
                if (!used[number]) {
                    // Check if k-th permutation is in the group starting with this number
                    if (k > factorial) {
                        // k-th permutation is not in this group, skip to next group
                        k -= factorial;
                    } else {
                        // k-th permutation starts with this number
                        result.append(number);
                        used[number] = true;
                        break;
                    }
                }
            }
        }
      
        return result.toString();
    }
}

File: leetcode/0060-permutation-sequence/0060-permutation-sequence.js
/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */
var getPermutation = function(n, k) {
    let result = "";

    // 각 숫자의 사용 여부를 저장
    const isUsed = Array(n + 1).fill(false);

    // 순열을 한 자리씩 생성
    for (let position = 0; position < n; position++) {
        // 남은 자리의 팩토리얼 계산
        let factorial = 1;
        for (let i = 1; i < n - position; i++) {
            factorial *= i;
        }

        // 사용할 숫자 선택
        for (let digit = 1; digit <= n; digit++) {
            if (!isUsed[digit]) {
                if (k > factorial) {
                    // 현재 숫자로 시작하는 모든 순열 건너뛰기
                    k -= factorial;
                } else {
                    // 현재 숫자를 선택
                    result += digit;
                    isUsed[digit] = true;
                    break;
                }
            }
        }
    }

    return result;
};